package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/lua
+
+local site = require 'gluon.site'
+local uci = require('simple-uci').cursor()
+local util = require 'gluon.util'
+local wireless = require 'gluon.wireless'
+
+-- Create wifi roles if they do not exist yet
+local default_wifi_roles = {
+	band_2g = {},
+	band_5g = {},
+}
+
+-- enable client if it is not disabled
+if site.wifi24.ap() and not site.wifi24.ap.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_2g'], 'client')
+end
+
+if site.wifi5.ap() and not site.wifi5.ap.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_5g'], 'client')
+end
+
+-- enable mesh if it is not disabled
+if site.wifi24.mesh() and not site.wifi24.mesh.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_2g'], 'mesh')
+end
+if site.wifi5.mesh() and not site.wifi5.mesh.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_5g'], 'mesh')
+end
+
+local role_iface_names = {
+	client = 'client_radio',
+	mesh = 'mesh_radio',
+	private = 'wan_radio',
+}
+
+local function migrate_wifi_to_roles(band)
+	local band_roles = {}
+	local config_exists = false
+	wireless.foreach_radio(uci, function(radio)
+		if radio.band ~= band then
+			return
+		end
+		local radio_name = radio['.name']
+		local suffix = radio_name:match('^radio(%d+)$')
+
+		for role, iface_prefix in pairs(role_iface_names) do
+			local iface_name = iface_prefix .. suffix
+			if uci:get('wireless', iface_prefix .. suffix) then
+				config_exists = true
+				local role_disabled = uci:get('wireless', iface_prefix .. suffix, 'disabled') or 0
+				if role_disabled == '0' then
+					util.add_to_set(band_roles, role)
+					if role == 'private' then
+						local ssid = uci:get('wireless', iface_name, 'ssid')
+						local key = uci:get('wireless', iface_name, 'key')
+						local encryption = uci:get('wireless', iface_name, 'encryption')
+						local mfp = uci:get('wireless', iface_name, 'ieee80211w')
+						uci:set('gluon', 'wireless', 'private_ssid', ssid)
+						uci:set('gluon', 'wireless', 'private_key', key)
+						uci:set('gluon', 'wireless', 'private_encryption', encryption)
+						uci:set('gluon', 'wireless', 'private_mfp', mfp)
+					end
+				end
+			end
+		end
+	end)
+
+	-- if no config exists for this band, we apply the default
+	if not config_exists then
+		band_roles = default_wifi_roles["band_" .. band]
+	end
+
+	uci:section('gluon', 'wireless_band', "band_" .. band, {
+		role = band_roles,
+	})
+end
+
+for band_name, _ in pairs(default_wifi_roles) do
+	local raw_band = band_name:gsub("^band_", "")
+	if not uci:get('gluon', band_name) and wireless.device_uses_band(uci, raw_band) then
```

Comment:
Hmm, I wonder if `device_uses_band` is the right choice here - it might be a good idea generate band configs for all bands that *could* be supported, allow switching the band when there is a single multi-band radio.